### PR TITLE
simple PyLmod implementation using the StellarGradeBook class almost unmodified

### PR DESCRIPTION
**Pull request by pwilkins**
This PR is response to @Ferdi's request to do the minimal version for 0.1.  It is a WIP because there are still some changes to bring it to the same level as PR#1.  This can be tested by running the exercise scripts in https://github.com/mitocw/grades2stellar/tree/features/pylmod_refactoring

### CODE REVIEW

File: pylmod/Client.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
+"""
```
Comment (carsongee) [inline pylmod/Client.py]: This file needs to be all lower case

**Comment by carsongee** [comment]
It would be nice if you had a paired branch of grades2stellar you could point me too to go along with this

File: pylmod/stellargradebook.py [hunk 2]
```diff
@@ -0,0 +1,555 @@
...
+
+    def rest_action_actual(self, fn, url, **kwargs):
+        """Routine to do low-level REST operation"""
+        log.info('Running request to %s' % url)
+        r = fn(url, timeout=self.TIMEOUT, verify=False, **kwargs)
+        try:
+            retdat = json.loads(r.content)
+        except Exception, err:
+            if (
+                    self.do_touchstone and
```
Comment (carsongee) [inline pylmod/stellargradebook.py]: this method doesn't exist anymore

File: pylmod/stellargradebook.py [hunk 3]
```diff
@@ -0,0 +1,555 @@
...
+    sg.spreadsheet2gradebook(datafn)
+
+    """
+
+    URLBASE = 'https://learning-modules-test.mit.edu:8443/service/gradebook'
+
+    GETS = {'academicterms': '',
+            'academicterm': '/{termCode}',
+            'gradebook': '?uuid={uuid}',
+    }
```
Comment (carsongee) [inline pylmod/stellargradebook.py]: bad indent

File: pylmod/stellargradebook.py [hunk 4]
```diff
@@ -0,0 +1,555 @@
...
+            "[StellarGradeBook] rest_action failure: exceed max retries"
+        )
+
+    def rest_action_actual(self, fn, url, **kwargs):
+        """Routine to do low-level REST operation"""
+        log.info('Running request to %s' % url)
+        r = fn(url, timeout=self.TIMEOUT, verify=False, **kwargs)
+        try:
+            retdat = json.loads(r.content)
+        except Exception, err:
```
Comment (carsongee) [inline pylmod/stellargradebook.py]: `err` is unused and should be here if you don't use it.

File: pylmod/Client.py (hunk 1 above)
Comment (pwilkins) [inline pylmod/Client.py]: oops.  corrected.

**Comment by carsongee** [comment]
None of this works for me.  I am getting stack traces from the LMOD service regardless of what call I make:

```
<h2>Internal error</h2>
<p/>

org.springframework.beans.TypeMismatchException: Failed to convert value of type 'java.lang.String' to required type 'java.lang.Long'; nested exception is org.springframework.core.convert.ConversionFailedException: Unable to convert value "None" from type 'java.lang.String' to type 'java.lang.Long'; nested exception is java.lang.NumberFormatException: For input string: "None"
	at org.springframework.beans.SimpleTypeConverter.convertIfNecessary(SimpleTypeConverter.java:55)
	at org.springframework.validation.DataBinder.convertIfNecessary(DataBinder.java:534)
	at org.springframework.web.bind.annotation.support.HandlerMethodInvoker.resolvePathVariable(HandlerMethodInvoker.java:705)
	at org.springframework.web.bind.annotation.support.HandlerMethodInvoker.resolveHandlerArguments(HandlerMethodInvoker.java:352)
	at org.springframework.web.bind.annotation.support.HandlerMethodInvoker.invokeHandlerMethod(HandlerMethodInvoker.java:171)
	at org.springframework.web.servlet.mvc.annotation.AnnotationMethodHandlerAdapter.invokeHandlerMethod(AnnotationMethodHandlerAdapter.java:426)
	at org.springframework.web.servlet.mvc.annotation.AnnotationMethodHandlerAdapter.handle(AnnotationMethodHandlerAdapter.java:414)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:790)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:719)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:644)
	at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:549)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:621)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:728)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:305)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:210)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:368)
	at org.springframework.security.web.authentication.switchuser.SwitchUserFilter.doFilter(SwitchUserFilter.java:178)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:380)
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.invoke(FilterSecurityInterceptor.java:109)
	at org.springframework.security.web.access.intercept.FilterSecurityInterceptor.doFilter(FilterSecurityInterceptor.java:83)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:380)
	at org.springframework.security.web.access.ExceptionTranslationFilter.doFilter(ExceptionTranslationFilter.java:97)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:380)
	at org.springframework.security.web.session.SessionManagementFilter.doFilter(SessionManagementFilter.java:100)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:380)
	at org.springframework.security.web.authentication.AnonymousAuthenticationFilter.doFilter(AnonymousAuthenticationFilter.java:78)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:380)
	at org.springframework.security.web.servletapi.SecurityContextHolderAwareRequestFilter.doFilter(SecurityContextHolderAwareRequestFilter.java:54)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:380)
	at org.springframework.security.web.savedrequest.RequestCacheAwareFilter.doFilter(RequestCacheAwareFilter.java:35)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:380)
	at edu.mit.gradebookng.services.security.TouchstoneAuthenticationFilter.doFilter(TouchstoneAuthenticationFilter.java:215)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:380)
	at org.springframework.security.web.context.SecurityContextPersistenceFilter.doFilter(SecurityContextPersistenceFilter.java:79)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:380)
	at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:169)
	at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:237)
	at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:167)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:243)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:210)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:222)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:123)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:472)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:171)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:99)
	at org.apache.catalina.valves.AccessLogValve.invoke(AccessLogValve.java:936)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:118)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:407)
	at org.apache.coyote.ajp.AjpProcessor.process(AjpProcessor.java:200)
	at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:589)
	at org.apache.tomcat.util.net.JIoEndpoint$SocketProcessor.run(JIoEndpoint.java:310)
	at java.util.concurrent.ThreadPoolExecutor$Worker.runTask(ThreadPoolExecutor.java:886)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:908)
	at java.lang.Thread.run(Thread.java:662)
Caused by: org.springframework.core.convert.ConversionFailedException: Unable to convert value "None" from type 'java.lang.String' to type 'java.lang.Long'; nested exception is java.lang.NumberFormatException: For input string: "None"
	at org.springframework.core.convert.support.ConversionUtils.invokeConverter(ConversionUtils.java:42)
	at org.springframework.core.convert.support.GenericConversionService.convert(GenericConversionService.java:184)
	at org.springframework.beans.TypeConverterDelegate.convertIfNecessary(TypeConverterDelegate.java:141)
	at org.springframework.beans.TypeConverterDelegate.convertIfNecessary(TypeConverterDelegate.java:93)
	at org.springframework.beans.SimpleTypeConverter.convertIfNecessary(SimpleTypeConverter.java:49)
	... 53 more
Caused by: java.lang.NumberFormatException: For input string: "None"
	at java.lang.NumberFormatException.forInputString(NumberFormatException.java:48)
	at java.lang.Long.parseLong(Long.java:410)
	at java.lang.Long.valueOf(Long.java:525)
	at org.springframework.util.NumberUtils.parseNumber(NumberUtils.java:158)
	at org.springframework.core.convert.support.StringToNumberConverterFactory$StringToNumber.convert(StringToNumberConverterFactory.java:59)
	at org.springframework.core.convert.support.StringToNumberConverterFactory$StringToNumber.convert(StringToNumberConverterFactory.java:1)
	at org.springframework.core.convert.support.GenericConversionService$ConverterFactoryAdapter.convert(GenericConversionService.java:528)
	at org.springframework.core.convert.support.ConversionUtils.invokeConverter(ConversionUtils.java:36)
	... 57 more

Cookies:
```

**Comment by pwilkins** [comment]
I have updated the description to include the Grades2Stellar branch to use for testing.

File: pylmod/stellargradebook.py [hunk 5]
```diff
@@ -0,0 +1,555 @@
...
+    gradebookid = None
+
+    # todo.do we want to pass gbuuid as param in __init__?
+    def __init__(self, cert, urlbase=None, gbuuid=None):
+        """
+        Initialize StellarGradeBook instance.
+
+          - urlbase:    URL base for gradebook API (defaults to self.URLBASE)
+            (still needs certs); default False
+          - gbuuid:     gradebook UUID (eg STELLAR:/project/mitxdemosite)
```
Comment (carsongee) [inline pylmod/stellargradebook.py]: You aren't adding this param to self here.  I think this is why I am getting 500's

Comment (carsongee) [inline pylmod/stellargradebook.py]: yep, so adding:
```python
if gbuuid is not None:
            self.gradebookid = self.get_gradebook_id(gbuuid)
```
to __init__ fixed my 500

**Comment by carsongee** [comment]
Be sure to clean up pep8 and pylint issues:
```
pep8 pylmod
pylmod/Client.py:9:1: E302 expected 2 blank lines, found 1
pylmod/__init__.py:5:18: E261 at least two spaces before inline comment
pylmod/stellargradebook.py:46:9: E124 closing bracket does not match visual indentation
pylmod/stellargradebook.py:108:24: E702 multiple statements on one line (semicolon)
pylmod/stellargradebook.py:341:9: E124 closing bracket does not match visual indentation
pylmod/stellargradebook.py:470:37: E131 continuation line unaligned for hanging indent
pylmod/stellargradebook.py:472:80: E501 line too long (80 > 79 characters)
```

```
pylint pylmod
pylint pylmod
No config file found, using default configuration
************* Module pylmod
E:  7,21: Undefined variable name 'StellarGradeBook' in __all__ (undefined-all-variable)
************* Module pylmod.Client
W:  5, 0: Relative import 'stellargradebook', should be 'pylmod.stellargradebook' (relative-import)
C:  7, 0: Invalid constant name "log" (invalid-name)
************* Module pylmod.stellargradebook
C: 46, 0: Wrong continued indentation.
        }
        ^  | (bad-continuation)
C:283, 0: Wrong hanging indentation before block.
        self, name, shortname, weight,
        ^   | (bad-continuation)
C:284, 0: Wrong hanging indentation before block.
        maxpoints, duedatestr, gradebookid='',
        ^   | (bad-continuation)
C:285, 0: Wrong hanging indentation before block.
        **kwargs
        ^   | (bad-continuation)
C:326, 0: Wrong hanging indentation before block.
        self, assignmentid, studentid, gradeval, gradebookid='', **kwargs
        ^   | (bad-continuation)
C:341, 0: Wrong continued indentation.
        }
        ^           | (bad-continuation)
C:380, 0: Wrong hanging indentation before block.
        self, datafn, create_assignments=True,
        ^   | (bad-continuation)
C:381, 0: Wrong hanging indentation before block.
        email_field=None, single=False
        ^   | (bad-continuation)
C:434, 0: Wrong hanging indentation before block.
        self, creader, create_assignments, email_field, NonAssignmentFields
        ^   | (bad-continuation)
C:470, 0: Wrong hanging indentation before block.
                                    'assignmentId' not in r.get('data')
                            |   |   ^ (bad-continuation)
C:515, 0: Wrong hanging indentation before block.
        self, creader, create_assignments, email_field, NonAssignmentFields
        ^   | (bad-continuation)
C:  1, 0: Missing module docstring (missing-docstring)
C: 15, 0: Invalid constant name "log" (invalid-name)
C: 68,12: Invalid attribute name "URLBASE" (invalid-name)
W: 76, 8: Specify string format arguments as logging function parameters (logging-not-lazy)
C: 78, 4: Invalid argument name "fn" (invalid-name)
W: 86,16: Specify string format arguments as logging function parameters (logging-not-lazy)
W: 92,16: Specify string format arguments as logging function parameters (logging-not-lazy)
C:101, 4: Invalid argument name "fn" (invalid-name)
W:103, 8: Specify string format arguments as logging function parameters (logging-not-lazy)
C:104, 8: Invalid variable name "r" (invalid-name)
F:108,12: Unable to import 'ipdb' (import-error)
C:108,25: More than one statement on a single line (multiple-statements)
E:110,20: Instance of 'StellarGradeBook' has no 'do_touchstone' member (no-member)
E:114,16: Instance of 'StellarGradeBook' has no 'init_touchstone' member (no-member)
W:107,26: Unused variable 'err' (unused-variable)
C:160, 8: Invalid variable name "gb" (invalid-name)
W:214,12: Redefining built-in 'map' (redefined-builtin)
C:168, 4: Invalid argument name "sectionName" (invalid-name)
C:220,12: Invalid argument name "x" (invalid-name)
C:220,12: Missing function docstring (missing-docstring)
W:196,21: Unused variable 'sec' (unused-variable)
C:230, 4: Invalid argument name "sectionName" (invalid-name)
C:230, 4: Missing method docstring (missing-docstring)
R:282, 4: Too many arguments (7/5) (too-many-arguments)
W:299, 8: Specify string format arguments as logging function parameters (logging-not-lazy)
W:301, 8: Specify string format arguments as logging function parameters (logging-not-lazy)
C:320,12: Invalid variable name "a" (invalid-name)
W:343, 8: Specify string format arguments as logging function parameters (logging-not-lazy)
C:374,12: Invalid variable name "s" (invalid-name)
C:406, 8: Invalid variable name "NonAssignmentFields" (invalid-name)
C:416,12: Invalid variable name "fp" (invalid-name)
C:418,12: Invalid variable name "fp" (invalid-name)
C:425,12: Invalid variable name "r" (invalid-name)
C:427,12: Invalid variable name "r" (invalid-name)
C:433, 4: Invalid argument name "NonAssignmentFields" (invalid-name)
R:433, 4: Too many local variables (25/15) (too-many-locals)
W:450,16: Specify string format arguments as logging function parameters (logging-not-lazy)
C:465,24: Invalid variable name "r" (invalid-name)
W:472,28: Specify string format arguments as logging function parameters (logging-not-lazy)
W:480,20: Specify string format arguments as logging function parameters (logging-not-lazy)
C:484,16: Invalid variable name "ok" (invalid-name)
W:487,23: Catching too general exception Exception (broad-except)
W:488,20: Specify string format arguments as logging function parameters (logging-not-lazy)
C:492,20: Invalid variable name "ok" (invalid-name)
W:501, 8: Specify string format arguments as logging function parameters (logging-not-lazy)
C:506, 8: Invalid variable name "r" (invalid-name)
C:507, 8: Invalid variable name "dt" (invalid-name)
W:508, 8: Specify string format arguments as logging function parameters (logging-not-lazy)
W:458,25: Unused variable 'assignment' (unused-variable)
W:434,23: Unused argument 'create_assignments' (unused-argument)
W:448,17: Unused variable 'student' (unused-variable)
C:514, 4: Invalid argument name "NonAssignmentFields" (invalid-name)
R:514, 4: Too many local variables (20/15) (too-many-locals)
C:536,24: Invalid variable name "r" (invalid-name)
W:547,24: Specify string format arguments as logging function parameters (logging-not-lazy)
W:552,16: Specify string format arguments as logging function parameters (logging-not-lazy)
C:557,16: Invalid variable name "r" (invalid-name)
W:530,25: Unused variable 'assignment' (unused-variable)
W:515,23: Unused argument 'create_assignments' (unused-argument)
W:525,17: Unused variable 'student' (unused-variable)

Global evaluation
-----------------
Your code has been rated at 6.49/10
```

**Comment by carsongee** [comment]
Verified against edx-platform with the changes suggested above, let me know when you are ready for round whatever we are at.

Also, to point platform at the g2xserver you can run it locally with just `python grades2stellar/g2s_xserver.py` and then in your devstack add:

```
from .devstack import * # pylint: disable=wildcard-import, unused-wildcard-import

FEATURES['REMOTE_GRADEBOOK_URL'] = 'http://<your_ip_or_hostname>:8104'
```

to `lms/envs/private.py` in the edx-platform directory.  Then you can load the devops course in via studio (or command line, or sysadmin dash) since it is configured to use the demo gradebook.

File: pylmod/stellargradebook.py (hunk 4 above)
Comment (pwilkins) [inline pylmod/stellargradebook.py]: removed.

File: pylmod/stellargradebook.py (hunk 5 above)
Comment (pwilkins) [inline pylmod/stellargradebook.py]: fixed

**Comment by pdpinch** [comment]
Just to write down what we discussed yesterday, all those "Invalid variable name" pylint errors come from the names being too short. I don't know what all of pylint's requirements are, but one- or two-letter names are too short. 

It looks like the easiest way to find out the specific meaning of an error is to do something like:
```
pylint --help-msg=invalid-name
```
Although I have to say the standard message for ```invalid-name``` leaves something to be desired. 

Carson, do you have an opinion about the other errors? I would think the missing docstrings are the biggest issue.

**Comment by carsongee** [comment]
The names don't really bother me that much since they just make it hard to read the code.  I'm generally more concerned with the programming runtime issues like unused variables that take up memory, useless strings, and overly broad exceptions which make debugging issues hard.  After those big ones, doc strings are really important (following the google style guide via a sphinx plugin: https://pypi.python.org/pypi/sphinxcontrib-napoleon), and then variable names should be more descriptive than a single letter or two will allow.

File: setup.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+README = open('README.rst').read()
+
+setup(
+    name='pylmod',
+    version=VERSION,
+    license='BSD',
+    author='MIT ODL Engineering',
+    author_email='odl-engineering@mit.edu',
+    url="http://github.com/mitodl/pylmod",
+    description="PyLmod provides Python Implementation of MIT Learning Modules",
```
Comment (carsongee) [inline setup.py]: `python setup.py test --flakes --pep8` fails here, looks like you have 1 too many characters

**Comment by carsongee** [comment]
Alright, other than that last little comment, and the small fix on the other side, I'm :bullettrain_side:

**Comment by carsongee** [comment]
Oh boy, merge time

**Pull request merged**